### PR TITLE
Add proto message builders

### PR DESCRIPTION
**Pull request by morganlaw**
PR containing just the proto message builders. To compile the .proto file, you can either use protoc or run `mvn clean package` which will compile it

### CODE REVIEW

File: callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionProtoMessageBuilder.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  private final ExecutionContextProtoMessageBuilder executionContextProtoMessageBuilder;
+  private final MessageContextProtoMessageBuilder messageContextProtoMessageBuilder;
+
+  /**
+   * @param executionContextProtoMessageBuilder Builder for ExecutionContext Protocol Buffer
+   *     Messages.
+   * @param messageContextProtoMessageBuilder Builder for MessageContext Protocol Buffer Messages.
+   */
+  public ExecutionProtoMessageBuilder(
```
Comment (cazhou3) [inline callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionProtoMessageBuilder.java:35]: What are your thoughts on making the build methods of `ExecutionContextProtoMessageBuilder` and `MessageContextProtoMessageBuilder` static, since each instance is the same (no instance variables in these classes)? Then this constructor isn't needed.

File: callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilder.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+  /**
+   * Gets the appropriate FlowType enumerator from the {@link ExecutionContext}. Used to indicate an
+   * error or request flow.
+   *
+   * @param executionContext {@link ExecutionContext} from which to extract the flow type.
+   * @return ExecutionContext.FlowType enumerator value
+   */
+  private Execute.ExecutionContext.FlowType getFlowType(ExecutionContext executionContext) {
+    if (executionContext.isErrorFlow()) {
```
Comment (cazhou3) [inline callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilder.java:61]: Nit: could we use if/else if/else here?

File: callout/src/main/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilder.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  /**
+   * Builds the FlowMapValue for the Flow Variable map in the Message Protocol Buffer Message.
+   *
+   * @param value Either a String or FlowInfo object to store in the Flow Variable map.
+   * @return FlowMapValue to be stored in the Flow Variable map in the Message Protocol Buffer
+   *     Message.
+   */
+  private Execute.Message.FlowMapValue buildFlowMapValue(Object value) {
```
Comment (cazhou3) [inline callout/src/main/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilder.java:92]: Apologies if I missed this, but where is this used?

File: callout/pom.xml [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            <groupId>com.apigee.edge</groupId>
+            <artifactId>expressions</artifactId>
+            <version>1.0.0</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.auth</groupId>
+            <artifactId>google-auth-library-oauth2-http</artifactId>
+            <version>0.21.1</version>
+        </dependency>
+<!--        <dependency>-->
```
Comment (cazhou3) [inline callout/pom.xml:40]: Is this unused? If so, we can delete it.

File: callout/src/main/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilder.java (hunk 3 above)
Comment (morganlaw) [reply callout/src/main/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilder.java:92]: It's not used right now, but I thought I should just leave that in there to be used when we can actually access the map directly and want to store values into the proto map

File: callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilder.java (hunk 2 above)
Comment (morganlaw) [reply callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilder.java:61]: sure. i wonder if there's a style guide for situations like this (where the statement within the conditionals is a return) and if making it if/else if/else is more readable?

File: callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionProtoMessageBuilder.java (hunk 1 above)
Comment (morganlaw) [reply callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionProtoMessageBuilder.java:35]: makes sense. changed

File: callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilder.java (hunk 2 above)
Comment (cazhou3) [reply callout/src/main/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilder.java:64]: Looks like there's no standard to scenarios like this, so up to you if you want to change it or not!

File: callout/pom.xml [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+            <version>1.7.30</version>
+        </dependency>
+        <dependency>
+            <groupId>org.powermock</groupId>
+            <artifactId>powermock-api-mockito2</artifactId>
```
Comment (mju) [inline callout/pom.xml:64]: We most likely don't need ot use power mock.

File: proto/execution.proto [hunk 6]
```diff
@@ -94,7 +96,7 @@ message ExecutionContext {
     string sub_category = 2;
     string name = 3;
     string reason = 4;
-    map<string, google.protobuf.Any> attributes = 5;
+    map<string, bytes> attributes = 5;
```
Comment (mju) [inline proto/execution.proto:99]: why this change?

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+import org.mockito.Spy;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertEquals;
```
Comment (mju) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java:32]: reformat this file using google-java-format

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+
+public class ExecutionContextProtoMessageBuilderTest {
+
+  private final String ATTRIBUTES_KEY1 = "key1";
```
Comment (mju) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java:38]: private static final

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+  private final String ATTRIBUTES_KEY2 = "key2";
+  private final String ATTRIBUTES_VALUE1 = "value1";
+  private final String ATTRIBUTES_VALUE2 = "value2";
+  private final String FAULT_NAME = "faultname";
+  private final String FAULT_REASON = "faultreason";
+  private final String FAULT_SUB_CATEGORY = "faultsubcategory";
+  Fault mockFault1 = mock(Fault.class);
```
Comment (mju) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java:45]: private

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+  Fault mockFault2 = mock(Fault.class);
```
Comment (mju) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java:46]: I think you can do @Mock mockFault2;

@Mock takes effect when the following in used in @Before

    MockitoAnnotations.initMocks(this);

You call openMocks(this).  You can see how the two differ.  Maybe one covers the other.

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionResultProtoMessageBuilderTest.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+import com.apigee.flow.execution.ExecutionResult;
+import com.google.apigee.Execute;
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertEquals;
```
Comment (mju) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionResultProtoMessageBuilderTest.java:27]: reformat this file.

https://github.com/google/google-java-format#intellij-android-studio-and-other-jetbrains-ides 

Read the last paragraph.  There is something about import ordering.

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionResultProtoMessageBuilderTest.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+public class ExecutionResultProtoMessageBuilderTest {
+
+  private final String KEY1 = "key1";
```
Comment (mju) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionResultProtoMessageBuilderTest.java:31]: private static final

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilderTest.java [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+import java.util.Map;
+import java.util.Set;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class MessageContextProtoMessageBuilderTest {
+
+  private final String CONTENT = "content1";
```
Comment (mju) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilderTest.java:38]: ditto

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java (hunk 10 above)
Comment (morganlaw) [reply callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java:46]: https://www.javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/MockitoAnnotations.html

initMocks has been deprecated and replaced with openMocks in mockito 3.4

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+  @Spy private FakeExecutionContext executionContext;
+
+  @Before
+  public void init() {
+    MockitoAnnotations.openMocks(this);
+
+    doReturn(Fault.Category.Messaging).when(mockFault1).getCategory();
```
Comment (cazhou3) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java:53]: Could you make a fake implementation for `Fault`? All the `when` statements are hard to parse at first glance.

You might want to look into [AutoValue](https://github.com/google/auto/blob/master/value/userguide/index.md) as well.

File: callout/src/main/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilder.java [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  /**
+   * Builds the FlowMapValue for the Flow Variable map in the Message Protocol Buffer Message.
+   *
+   * @param value Either a String or FlowInfo object to store in the Flow Variable map.
+   * @return FlowMapValue to be stored in the Flow Variable map in the Message Protocol Buffer
+   *     Message.
+   */
+  private static Execute.Message.FlowMapValue buildFlowMapValue(Object value) {
```
[the remark this answers is not in the document]
Comment (morganlaw) [reply callout/src/main/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilder.java:92]: christine called this out earlier as well, it's not used right now since we can't extract the map directly yet, but i just left that in there for use in the future once we can do that.

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java (hunk 14 above)
Comment (morganlaw) [reply callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java]: I don't think(?) I can make a fake implementation for Fault since it's not an interface and extending it would require calling the superclass constructor where is where our issue is. I did play around with AutoValue, but I'm not sure I can use it here for the same reason.

File: callout/src/main/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilder.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (mju) [reply callout/src/main/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilder.java:91]: comment on this and remove private so it can be tested?

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java (hunk 14 above)
Comment (cazhou3) [reply callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java]: sg, feel free to leave in the two faults and the when statements if you feel like it will help test your code better.

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+  @Test
+  public void testBuildExecutionContextProtoWithMultipleFaults() {
+    executionContext.addFault(mockFault);
+    executionContext.addFault(mockFault);
+    doReturn(Fault.Category.Messaging).when(mockFault).getCategory();
+    Execute.ExecutionContext proto =
+        ExecutionContextProtoMessageBuilder.buildExecutionContextProto(executionContext);
+
+    assertEquals(2, proto.getFaultsCount());
+    assertEquals(
```
Comment (cazhou3) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/ExecutionContextProtoMessageBuilderTest.java:126]: nit: could you try to reformat this file with `google-java-format`? I think lines 126 and 132 are the same but they are line separated differently

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilderTest.java [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+
+  abstract static class FakeMessageContext implements MessageContext {
+    private Message proxyRequestMessage;
+    private Message targetRequestMessage;
+    private Message proxyResponseMessage;
+    private Message targetResponseMessage;
+    private Message errorMessage;
+
+    @Override
+    public Message getMessage(FlowContext var1) {
```
Comment (cazhou3) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilderTest.java:195]: nit: we can use more descriptive variable names in this fake

File: callout/src/test/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilderTest.java [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+    @Override
+    public Set<String> getHeaderNames() {
+      if (headers == null) {
+        return null;
+      }
+      return headers.keySet();
+    }
+
+    @Override
+    public List<String> getHeaders(String var1) {
```
Comment (cazhou3) [inline callout/src/test/java/com/google/apigee/ProtoMessageBuilders/MessageContextProtoMessageBuilderTest.java:261]: same as above
